### PR TITLE
Add feature type and feature name handling

### PR DESCRIPTION
**Pull request by madskonradsen**
Right now, the DMatrix is internal to the booster, but it might make sense to simply let people pass a DMatrix directly to the classifier/regressor?

### CODE REVIEW

**Comment by mdabros** [comment]
> Right now, the DMatrix is internal to the booster, but it might make sense to simply let people pass a DMatrix directly to the classifier/regressor?

Yes, I think that would be a good idea to provide more control for more advanced use cases. We can keep the float array overloads for the more basic/easy cases.

**Comment by mdabros** [comment]
I will try to review the PR this week.

**Comment by madskonradsen** [comment]
Added the overloads for DMatrix and basic tests just to make use of them

File: src/XGBoostSharp/lib/DMatrix.cs [hunk 1]
```diff
@@ -68,6 +68,51 @@ void SetFloatInfo(string field, float[] floatInfo)
         ThrowIfError(output);
     }
 
+    public void SetFeatureNames(string[] featureNames)
+    {
+        SetFeatureInfo(featureNames, "feature_name");
```
Comment (mdabros) [inline src/XGBoostSharp/lib/DMatrix.cs:73]: @madskonradsen Let's move the hardcoded strings to a `Fields` class, so we only define them in one place. like this:

```csharp
namespace XGBoostSharp.lib;

public static class Fields
{
#pragma warning disable IDE1006 // Naming Styles
    public const string label = nameof(label);
    public const string feature_name = nameof(feature_name);
    public const string feature_type = nameof(feature_type);
#pragma warning restore IDE1006 // Naming Styles
}
```

File: src/XGBoostSharp/lib/DMatrix.cs [hunk 2]
```diff
@@ -68,6 +68,51 @@ void SetFloatInfo(string field, float[] floatInfo)
...
+    {
+        return GetFeatureInfo("feature_name");
+    }
+
+    public void SetFeatureTypes(string[] featureTypes)
+    {
+        SetFeatureInfo(featureTypes, "feature_type");
+    }
+
+    public string[] GetFeatureTypes()
```
Comment (mdabros) [inline src/XGBoostSharp/lib/DMatrix.cs:86]: @madskonradsen For these one-liner methods please use:
```csharp
    public string[] GetFeatureTypes() => GetFeatureInfo(Fields.feature_type);
```

File: src/XGBoostSharp/XGBClassifier.cs [hunk 3]
```diff
@@ -183,17 +194,38 @@ public float[] Predict(float[][] data)
...
+    /// <summary>
+    ///   Predict using the gradient boosted model
+    /// </summary>
+    /// <param name="dMatrix">
+    ///   DMatrix to do predictions on
+    /// </param>
+    /// <returns>
+    ///   Predictions
+    /// </returns>
+    public float[] Predict(DMatrix dMatrix)
```
Comment (mdabros) [inline src/XGBoostSharp/XGBClassifier.cs:206]: @madskonradsen Where possible, we might as well reuse and call the new DMatrix overload methods from the old float based methods. For both `Predict` and `Fit` methods, and for both the regressor and the classifier implementations.

File: src/XGBoostSharp.Tests/DMatrixTest.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+{
+    readonly float[][] m_dataTrain = [[1, 2, 3, 4]];
+    readonly float[] m_labelsTrain = [1];
+
+    [TestMethod]
+    public void DMatrix_SetFeatureNames()
+    {
+        string[] featureNames = ["f1", "f2", "f3", "f4"];
+
+        var sut = new DMatrix(m_dataTrain, m_labelsTrain);
```
Comment (mdabros) [inline src/XGBoostSharp.Tests/DMatrixTest.cs:17]: @madskonradsen Remember using statement for the DMatrix:
```csharp
using var sut = new DMatrix(m_dataTrain, m_labelsTrain);
```
Same for the the other tests.

**Review by mdabros — CHANGES_REQUESTED** [review]
@madskonradsen I have added a few comments and suggestions - nothing major :). Besides that it looks good 👍 .

**Comment by madskonradsen** [comment]
Fair points, have fixed accordingly! :)

**Comment by mdabros** [comment]
> Fair points, have fixed accordingly! :)

Looks good. I will merge and release a new version with the additions. Thanks for the contribution!

**Pull request merged**